### PR TITLE
Search package: refactoring search module control

### PR DESCRIPTION
**Pull request by kangzj**
#### Changes proposed in this Pull Request:
Refactored the logic in search `Module_Control` and `REST_Controller` class. Added `update_status` and `update_instant_search_status` for easier calls

#### Jetpack product discussion
n/a

#### Does this pull request change what data or activity we track or use?
no

#### Testing instructions:
* Run `jetpack install --all`
* `(cd projects/packages/search && composer phpunit)`
* Ensure all tests pass

### CODE REVIEW

File: projects/packages/search/src/class-module-control.php [hunk 1]
```diff
@@ -148,23 +148,49 @@ public function deactivate() {
...
+			return true;
+		}
 		return update_option( self::SEARCH_MODULE_INSTANT_SEARCH_OPTION_KEY, false );
 	}
 
 	/**
 	 * Enable Instant Search Experience
 	 */
 	public function enable_instant_search() {
+		if ( ! $this->is_active() || $this->is_instant_search_enabled() ) {
```
Comment (jsnmoon) [inline projects/packages/search/src/class-module-control.php:179]: Shouldn't this return `false` for `! $this->is_active()` to indicate that the option was not updated? In other words, wouldn't this scenario be equivalent to a 400 error?

File: projects/packages/search/src/class-module-control.php [hunk 2]
```diff
@@ -148,23 +148,49 @@ public function deactivate() {
...
+		return $active ? $this->activate() : $this->deactivate();
+	}
+
 	/**
 	 * Disable Instant Search Experience
 	 */
 	public function disable_instant_search() {
+		if ( ! $this->is_instant_search_enabled() ) {
+			return true;
+		}
```
Comment (jsnmoon) [inline projects/packages/search/src/class-module-control.php:171]: Why is this check necessary? What's the benefit of bypassing the `update_option` call here?

**Review by jsnmoon — COMMENTED** [review]
Good refactor of `update_settings`! It's much more readable.

Left some questions inline, but tests pass without issue. Could you provide additional instructions on how to smoke test the actual API endpoints?

File: projects/packages/search/src/class-module-control.php (hunk 2 above)
Comment (kangzj) [reply projects/packages/search/src/class-module-control.php:171]: The only benefit is on some condition, we could use less write operations. But I guess you are right, no need to do so, the situation is pretty rare. No need to do this.

Comment (kangzj) [reply projects/packages/search/src/class-module-control.php:171]: removed in 78aae3f1e7

File: projects/packages/search/src/class-module-control.php (hunk 1 above)
Comment (kangzj) [reply projects/packages/search/src/class-module-control.php:179]: Good catch here. If the module is not active then we definitely should return false or an error.

**Comment by kangzj** [comment]
Thanks for the speedy review @jsnmoon 

> Could you provide additional instructions on how to smoke test the actual API endpoints?

You could find the testing instructions here: https://github.com/Automattic/jetpack/pull/21584

File: projects/packages/search/src/class-module-control.php (hunk 1 above)
Comment (kangzj) [reply projects/packages/search/src/class-module-control.php]: addressed in 3d57b30851

**Comment by kangzj** [comment]
@jsnmoon added proper errors in ef628b6. Thanks for the suggestion, which will save some future confusion. 👍

**Review by jsnmoon — DISMISSED** [review]
Works as described 👍 

(Needs a rebase)

**Review by jsnmoon — APPROVED** [review]

**Pull request merged**